system/jlib/jmetrics.hpp

```diff
@@ -0,0 +1,295 @@
+    virtual const std::string &queryName() const = 0;
+
+    /*
+     * Returns metric description
+     */
+    virtual const std::string &queryDescription() const = 0;
+
+    /*
+     * Returns the metric type.
+     */
+    virtual MetricType queryMetricType() const = 0;
+
+    /*
+     * Get report values
+     */
+    virtual void collect(MeasurementVector &mv) const = 0;
+};
+
+
+/*
+ * Concrete base class implementation of the IMetric interface. All metrics inherit
+ * from this class.
+*/
+class jlib_decl Metric : public IMetric
+{
+    public:
+        virtual ~Metric() = default;
+        const std::string &queryName() const override { return name; }
+        const std::string &queryDescription() const override { return description; }
+        MetricType queryMetricType() const override { return metricType; }
+
+    protected:
+        // No one should be able to create one of these
+        Metric(const char *_name, const char *_desc, MetricType _metricType) :
+                name{_name},
+                description{_desc},
+                metricType{_metricType} { }
+
+    protected:
+        std::string name;
+        std::string description;
+        MetricType metricType;
+};
+
+/*
+ * Metric used to count events. Count is a monotonically increasing value
+ */
+class jlib_decl CounterMetric : public Metric
+{
+    public:
+        CounterMetric(const char *name, const char *description) :
+                Metric{name, description, MetricType::METRICS_COUNTER}  { }
+        void inc(uint32_t val)
+        {
+            count.fetch_add(val);
+        }
+
+        void collect(MeasurementVector &mv) const override;
+
+    protected:
+        std::atomic<uint32_t> count{0};
+};
+
+
+/*
+ * Metric used to track the current state of some internal measurement.
+ */
+class jlib_decl GaugeMetric : public Metric {
+    public:
+        GaugeMetric(const char *name, const char *description) :
+            Metric{name, description, MetricType::METRICS_GAUGE}  { }
+
+        /*
+         * Update the value as indicated
+         */
+        void add(int32_t delta)
+        {
+            gaugeValue += delta;
+        }
+
+        /*
+         * Set the value
+         */
+        void set(int32_t val) { gaugeValue = val; }
+
+        /*
+         * Read the current value
+         */
+        int32_t queryValue() const { return gaugeValue; }
+
+        /*
+         * Get current gauge value
+         */
+        void collect(MeasurementVector &mv) const override;
+
+    protected:
+        std::atomic<int32_t> gaugeValue{0};
+};
+
+
+
+class jlib_decl MetricSink
+{
+    public:
+        virtual ~MetricSink() = default;
+        virtual void startCollection(MetricsReporter *pReporter) = 0;
+        virtual void stopCollection() = 0;
+        const std::string &queryName() const { return name; }
+        const std::string &queryType() const { return type; }
+
+    protected:
+        MetricSink(std::string _name, std::string _type) :
```

Comment:
not sure why std::string parameters are used here.  (The base classes don't have them.)  More normal is to pass const char *